### PR TITLE
Dm/feature/incremental2

### PR DESCRIPTION
**Pull request by dbmeneses**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/computation/queue/ReportSubmitter.java [hunk 1]
```diff
@@ -129,7 +137,7 @@ private ComponentDto createProject(DbSession dbSession, OrganizationDto organiza
     return componentUpdater.create(dbSession, newProject, userId);
   }
 
-  private CeTask submitReport(DbSession dbSession, InputStream reportInput, ComponentDto project) {
+  private CeTask submitReport(DbSession dbSession, InputStream reportInput, ComponentDto project, Map<String, String> characteristics) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/queue/ReportSubmitter.java]: MAJOR Remove this unused method parameter "characteristics". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1172)

File: server/sonar-server/src/main/java/org/sonar/server/ce/ws/SubmitAction.java [hunk 2]
```diff
@@ -95,6 +103,7 @@ public void handle(Request wsRequest, Response wsResponse) throws Exception {
     String projectKey = wsRequest.mandatoryParam(PARAM_PROJECT_KEY);
     String projectBranch = wsRequest.param(PARAM_PROJECT_BRANCH);
     String projectName = StringUtils.defaultIfBlank(wsRequest.param(PARAM_PROJECT_NAME), projectKey);
+    //String[] characteristics = StringUtils.(wsRequest.getParam(PARAM_ANALYSIS_CHARACTERISTICS).or(() -> ""), '=');
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/ce/ws/SubmitAction.java]: MAJOR This block of commented-out lines of code should be removed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues
* ![MAJOR][MAJOR] 6 major

Watch the comments in this conversation to review them.

#### 4 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [MyBatis.java#L134](https://github.com/SonarSource/sonarqube/blob/ad59928402713e1dc04b05e41e2f19668b262aea/server/sonar-db-dao/src/main/java/org/sonar/db/MyBatis.java#L134): This method has 101 lines, which is greater than the 100 lines authorized. Split it into smaller methods. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS138)
1. ![MAJOR][MAJOR] [v66](https://github.com/SonarSource/sonarqube/blob/ad59928402713e1dc04b05e41e2f19668b262aea/server/sonar-db-migration/src/main/java/org/sonar/server/platform/db/migration/version/v66): Add a 'package-info.java' file to document the 'v66' package [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [PersistFileSourcesStep.java#L76](https://github.com/SonarSource/sonarqube/blob/ad59928402713e1dc04b05e41e2f19668b262aea/server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/PersistFileSourcesStep.java#L76): Constructor has 8 parameters, which is greater than 7 authorized. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00107)
1. ![MAJOR][MAJOR] [validation](https://github.com/SonarSource/sonarqube/blob/ad59928402713e1dc04b05e41e2f19668b262aea/server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/validation): Add a 'package-info.java' file to document the 'validation' package [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1228)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/measure/MeasureRepositoryImpl.java [hunk 3]
```diff
@@ -21,8 +21,16 @@
 
 import com.google.common.base.Optional;
 import com.google.common.collect.SetMultimap;
+
+import java.util.ArrayList;
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/measure/MeasureRepositoryImpl.java]: MINOR Remove this unused import 'java.util.ArrayList'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/measure/MeasureRepositoryImpl.java [hunk 4]
```diff
@@ -21,8 +21,16 @@
...
+import java.util.Collection;
+import java.util.HashMap;
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/measure/MeasureRepositoryImpl.java]: MINOR Remove this unused import 'java.util.HashMap'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 8 issues
* ![MAJOR][MAJOR] 6 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

#### 4 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [MyBatis.java#L134](https://github.com/SonarSource/sonarqube/blob/84a02002baeb930cd6dfb8377c644f7dfecb86f5/server/sonar-db-dao/src/main/java/org/sonar/db/MyBatis.java#L134): This method has 101 lines, which is greater than the 100 lines authorized. Split it into smaller methods. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS138)
1. ![MAJOR][MAJOR] [v66](https://github.com/SonarSource/sonarqube/blob/84a02002baeb930cd6dfb8377c644f7dfecb86f5/server/sonar-db-migration/src/main/java/org/sonar/server/platform/db/migration/version/v66): Add a 'package-info.java' file to document the 'v66' package [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [PersistFileSourcesStep.java#L76](https://github.com/SonarSource/sonarqube/blob/84a02002baeb930cd6dfb8377c644f7dfecb86f5/server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/PersistFileSourcesStep.java#L76): Constructor has 8 parameters, which is greater than 7 authorized. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00107)
1. ![MAJOR][MAJOR] [validation](https://github.com/SonarSource/sonarqube/blob/84a02002baeb930cd6dfb8377c644f7dfecb86f5/server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/validation): Add a 'package-info.java' file to document the 'validation' package [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1228)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

#### 3 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [MyBatis.java#L135](https://github.com/SonarSource/sonarqube/blob/8bf06f38de6861a645d80b2a05defb1fe0fe7b84/server/sonar-db-dao/src/main/java/org/sonar/db/MyBatis.java#L135): This method has 102 lines, which is greater than the 100 lines authorized. Split it into smaller methods. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS138)
1. ![MAJOR][MAJOR] [PersistFileSourcesStep.java#L76](https://github.com/SonarSource/sonarqube/blob/8bf06f38de6861a645d80b2a05defb1fe0fe7b84/server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/PersistFileSourcesStep.java#L76): Constructor has 8 parameters, which is greater than 7 authorized. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00107)
1. ![MAJOR][MAJOR] [validation](https://github.com/SonarSource/sonarqube/blob/8bf06f38de6861a645d80b2a05defb1fe0fe7b84/server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/validation): Add a 'package-info.java' file to document the 'validation' package [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1228)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [MyBatis.java#L135](https://github.com/SonarSource/sonarqube/blob/d5d0a61c6d18b455de8c19828df1d73cc289a089/server/sonar-db-dao/src/main/java/org/sonar/db/MyBatis.java#L135): This method has 102 lines, which is greater than the 100 lines authorized. Split it into smaller methods. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS138)
1. ![MAJOR][MAJOR] [PersistFileSourcesStep.java#L76](https://github.com/SonarSource/sonarqube/blob/d5d0a61c6d18b455de8c19828df1d73cc289a089/server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/PersistFileSourcesStep.java#L76): Constructor has 8 parameters, which is greater than 7 authorized. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00107)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: server/sonar-db-core/src/main/resources/org/sonar/db/version/schema-h2.ddl [hunk 5]
```diff
@@ -610,11 +610,18 @@ CREATE TABLE "CE_ACTIVITY" (
...
+
 CREATE UNIQUE INDEX "CE_ACTIVITY_UUID" ON "CE_ACTIVITY" ("UUID");
 CREATE INDEX "CE_ACTIVITY_COMPONENT_UUID" ON "CE_ACTIVITY" ("COMPONENT_UUID");
 CREATE INDEX "CE_ACTIVITY_ISLASTKEY" ON "CE_ACTIVITY" ("IS_LAST_KEY");
 CREATE INDEX "CE_ACTIVITY_ISLAST_STATUS" ON "CE_ACTIVITY" ("IS_LAST", "STATUS");
 
+CREATE TABLE "CE_TASK_CHARACTERISTICS" (
+  "UUID" VARCHAR(40) NOT NULL PRIMARY KEY,
+  "TASK_UUID" VARCHAR(40) NOT NULL,
+  "KEE" VARCHAR(512) NOT NULL,
```
Comment (simonbrandhof) [inline server/sonar-db-core/src/main/resources/org/sonar/db/version/schema-h2.ddl]: no need for a so big key. 50 characters is good enough.

File: server/sonar-db-core/src/main/resources/org/sonar/db/version/schema-h2.ddl [hunk 6]
```diff
@@ -610,11 +610,18 @@ CREATE TABLE "CE_ACTIVITY" (
...
+  "TEXT_VALUE" VARCHAR(512)
```
Comment (simonbrandhof) [inline server/sonar-db-core/src/main/resources/org/sonar/db/version/schema-h2.ddl]: Usually we put the maximum supported length (4000) when it's above a few hundreds characters.

File: server/sonar-db-core/src/main/resources/org/sonar/db/version/schema-h2.ddl [hunk 7]
```diff
@@ -610,11 +610,18 @@ CREATE TABLE "CE_ACTIVITY" (
...
 );
+
 CREATE UNIQUE INDEX "CE_ACTIVITY_UUID" ON "CE_ACTIVITY" ("UUID");
 CREATE INDEX "CE_ACTIVITY_COMPONENT_UUID" ON "CE_ACTIVITY" ("COMPONENT_UUID");
 CREATE INDEX "CE_ACTIVITY_ISLASTKEY" ON "CE_ACTIVITY" ("IS_LAST_KEY");
 CREATE INDEX "CE_ACTIVITY_ISLAST_STATUS" ON "CE_ACTIVITY" ("IS_LAST", "STATUS");
 
+CREATE TABLE "CE_TASK_CHARACTERISTICS" (
+  "UUID" VARCHAR(40) NOT NULL PRIMARY KEY,
+  "TASK_UUID" VARCHAR(40) NOT NULL,
```
Comment (simonbrandhof) [inline server/sonar-db-core/src/main/resources/org/sonar/db/version/schema-h2.ddl]: An index is missing on TASK_UUID in order to quickly load the characteristics of a task (see the SQL request `selectTaskCharacteristics`).

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [MyBatis.java#L135](https://github.com/SonarSource/sonarqube/blob/40d483aefe8c8d600971ecf1ea2c86d775298a5a/server/sonar-db-dao/src/main/java/org/sonar/db/MyBatis.java#L135): This method has 102 lines, which is greater than the 100 lines authorized. Split it into smaller methods. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS138)
1. ![MAJOR][MAJOR] [PersistFileSourcesStep.java#L76](https://github.com/SonarSource/sonarqube/blob/40d483aefe8c8d600971ecf1ea2c86d775298a5a/server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/PersistFileSourcesStep.java#L76): Constructor has 8 parameters, which is greater than 7 authorized. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00107)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [MyBatis.java#L135](https://github.com/SonarSource/sonarqube/blob/38549b47aa7a7ff13e7bfa0621d2a279dd7f19e3/server/sonar-db-dao/src/main/java/org/sonar/db/MyBatis.java#L135): This method has 102 lines, which is greater than the 100 lines authorized. Split it into smaller methods. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS138)
1. ![MAJOR][MAJOR] [PersistFileSourcesStep.java#L76](https://github.com/SonarSource/sonarqube/blob/38549b47aa7a7ff13e7bfa0621d2a279dd7f19e3/server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/PersistFileSourcesStep.java#L76): Constructor has 8 parameters, which is greater than 7 authorized. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00107)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [MyBatis.java#L135](https://github.com/SonarSource/sonarqube/blob/709c6116e8ee9eafad79aeccbbde5690743727dd/server/sonar-db-dao/src/main/java/org/sonar/db/MyBatis.java#L135): This method has 102 lines, which is greater than the 100 lines authorized. Split it into smaller methods. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS138)
1. ![MAJOR][MAJOR] [PersistFileSourcesStep.java#L76](https://github.com/SonarSource/sonarqube/blob/709c6116e8ee9eafad79aeccbbde5690743727dd/server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/PersistFileSourcesStep.java#L76): Constructor has 8 parameters, which is greater than 7 authorized. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00107)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [MyBatis.java#L134](https://github.com/SonarSource/sonarqube/blob/021be7deae1b11c211a37073bcde571fe29c9a04/server/sonar-db-dao/src/main/java/org/sonar/db/MyBatis.java#L134): This method has 101 lines, which is greater than the 100 lines authorized. Split it into smaller methods. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS138)
1. ![MAJOR][MAJOR] [PersistFileSourcesStep.java#L76](https://github.com/SonarSource/sonarqube/blob/021be7deae1b11c211a37073bcde571fe29c9a04/server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/PersistFileSourcesStep.java#L76): Constructor has 8 parameters, which is greater than 7 authorized. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00107)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Pull request merged**